Mathlib/RingTheory/DedekindDomain/FiniteAdeleRing.lean

```diff
@@ -364,6 +370,122 @@ instance : Algebra (R_hat R K) (FiniteAdeleRing R K) where
+    -- use ϖ^d
+    refine ⟨ϖ^d.natAbs, pow_mem (mem_nonZeroDivisors_of_ne_zero hϖ0) _, ?_⟩
+    -- now manually translate the goal (an inequality in ℤₘ₀) to an inequality in ℤ
+    rw [mem_adicCompletionIntegers, algebraMap.coe_pow, map_mul, hd, map_pow,
+      valuedAdicCompletion_eq_valuation, valuation_eq_intValuationDef, hϖ, ← WithZero.coe_pow,
+      ← WithZero.coe_mul, WithZero.coe_le_one, ← toAdd_le, toAdd_mul, toAdd_ofAdd, toAdd_pow,
+      toAdd_ofAdd, toAdd_one,
+      show d.natAbs • (-1) = (d.natAbs : ℤ) • (-1) by simp only [nsmul_eq_mul,
+        Int.natCast_natAbs, smul_eq_mul],
+      ← Int.eq_natAbs_of_zero_le ha.le, smul_eq_mul]
+    -- and now it's easy
+    linarith
+
+open scoped DiscreteValuation
+
+variable {R K} in
+lemma clear_denominator (a : FiniteAdeleRing R K) :
+    ∃ (b : R⁰) (c : R_hat R K), a * ((b : R) : FiniteAdeleRing R K) = c := by
+  let S := {v | a v ∉ adicCompletionIntegers K v}
+  choose b hb h using clear_local_denominator (R := R) (K := K)
+  let p := ∏ᶠ v ∈ S, b v (a v)
+  have hp : p ∈ R⁰ := finprod_mem_induction (· ∈ R⁰) (one_mem _) (fun _ _ => mul_mem) <|
+    fun _ _ ↦ hb _ _
+  use ⟨p, hp⟩
+  rw [exists_finiteIntegralAdele_iff]
+  intro v
+  by_cases hv : a v ∈ adicCompletionIntegers K v
+  · exact mul_mem hv <| coe_mem_adicCompletionIntegers _ _
+  · change v ∈ S at hv
+    dsimp only
+    have pprod : p = b v (a v) * ∏ᶠ w ∈ S \ {v}, b w (a w) := by
+      rw [← finprod_mem_singleton (a := v) (f := fun v ↦ b v (a v)),
+        finprod_mem_mul_diff (singleton_subset_iff.2 hv) a.2]
+    rw [pprod]
+    push_cast
+    rw [← mul_assoc]
+    exact mul_mem (h v (a v)) <| coe_mem_adicCompletionIntegers _ _
+
+open scoped Pointwise
+
+theorem submodulesRingBasis : SubmodulesRingBasis
+    (fun (r : R⁰) ↦ Submodule.span (R_hat R K) {((r : R) : FiniteAdeleRing R K)}) where
+  inter i j := ⟨i * j, by
+    push_cast
+    simp only [le_inf_iff, Submodule.span_singleton_le_iff_mem, Submodule.mem_span_singleton]
+    exact ⟨⟨((j : R) : R_hat R K), by rw [mul_comm]; rfl⟩, ⟨((i : R) : R_hat R K), rfl⟩⟩⟩
+  leftMul a r := by
+    rcases clear_denominator a with ⟨b, c, h⟩
+    use r * b
+    rintro x ⟨m, hm, rfl⟩
+    simp only [Submonoid.coe_mul, SetLike.mem_coe] at hm
+    rw [Submodule.mem_span_singleton] at hm ⊢
+    rcases hm with ⟨n, rfl⟩
+    simp only [LinearMapClass.map_smul, DistribMulAction.toLinearMap_apply, smul_eq_mul]
+    use n * c
+    push_cast
+    rw [mul_left_comm, h, mul_comm _ (c : FiniteAdeleRing R K),
+      Algebra.smul_def', Algebra.smul_def', ← mul_assoc]
+    rfl
+  mul r := ⟨r, by
+    intro x hx
+    rw [mem_mul] at hx
+    rcases hx with ⟨a, ha, b, hb, rfl⟩
+    simp only [SetLike.mem_coe, Submodule.mem_span_singleton] at ha hb ⊢
+    rcases ha with ⟨m, rfl⟩
+    rcases hb with ⟨n, rfl⟩
+    use m * n * (r : R)
+    simp only [Algebra.smul_def', map_mul]
+    rw [mul_mul_mul_comm, mul_assoc]
+    rfl⟩
+
+instance : Nonempty (R⁰) := ⟨1, Submonoid.one_mem R⁰⟩
```

Comment:
This instance should be found via `One.instNonempty` right? Do we need this?
Even if we do, it probably doesn't belong here.

Reply:
You're right, it can be deleted.